File: DSA/21118_DSA_A04.py
# Author : Shubham (Roll No.: 21118)
# DSA Assignment 04 : Hashing Techniques

class Record():
    def __init__(self, num, name):
        self.ph_num, self.name = num, name

    def __str__(self):
        return "Phone Num: " + str(self.ph_num) + ", Name: " + str(self.name)


class HashTable():
    def __init__(self, D):
        self.MAX_SIZE = D
        self.table = [-1 for _ in range(D)]
        self.nxtIdx = [-1 for _ in range(D)]
        self.is_occu = [0 for _ in range(D)]

    def Print(self):
        i=0
        for entry in self.table:
            print(entry, end=" ")
            print(self.nxtIdx[i])
            i += 1
        print("--------------")

    def getHashVal(self, key):
        return key % self.MAX_SIZE

    def Search(self, key):
        idx = self.getHashVal(key)
        while self.table[idx] != -1:
            if self.table[idx].ph_num == key:
                self.table[idx].Print()
                return True
            idx = self.nxtIdx[idx]
        return False

    def insert(self, key, val):
        if (self.Search(key)):
            print("Record is already Present.")
            return
        idx = self.getHashVal(key)
        while self.is_occu[idx]:
            idx = (idx+1) % self.MAX_SIZE
        self.table[idx] = Record(key, val)
        self.is_occu[idx] = 1


def main():
    ht = HashTable(10)
    ht.insert(22, "A")
    ht.Print()
    ht.insert(33, "B")
    ht.Print()
    ht.insert(42, "C")
    ht.Print()
    ht.insert(32, "D")
    ht.Print()

main()

File: DSA/Assignment04/21118_DSA_A04.cpp
//============================================================================
// Name        : 21118_DSA_A04.cpp
// Author      : Shubham (Roll No: 21118)
//============================================================================

#include <iostream>
using namespace std;

const int MAX_SIZE = 10;

class Entry {
private:
	int key;
	string val;
public:
	Entry(int k = -1, string v = "$") {
		key = k;
		val = v;
	}
	void PrintEntry() {
		cout << key << " -> " << val << endl;
	}
	friend class HashTable;
};

int getSumOfSqrsOfDigits(int x) {
	int sum = 0;
	while (x != 0) {
		int d = x % 10;
		sum += (d * d);
		x /= 10;
	}
	return sum;
}

int getSumOfDigits(int x) {
	int sum = 0;
	while (x != 0) {
		int d = x % 10;
		sum += d;
		x /= 10;
	}
	return sum;
}

class HashTable {
private:
	Entry ht[MAX_SIZE];
public:
	HashTable() {

	}
	int getHashVal(int x) {
//		int sum_of_squares_of_dig = getSumOfSqrsOfDigits(x);
//		return (sum_of_squares_of_dig % MAX_SIZE);
		return (x % MAX_SIZE);
//		int sum_of_dig = getSumOfDigits(x);
//		return (sum_of_dig % MAX_SIZE);
	}
	int Insert(int key, string val) {
		Entry newEntry = Entry(key, val);
		int hash_val = getHashVal(key);
		int pos = hash_val, no_of_comp = 0;;
		while (ht[pos].key != -1) {
			no_of_comp++;
			pos = (pos+1) % MAX_SIZE;
			if (pos == hash_val) {
				cout << "Hash Table is Full.\n";
				return -1;
			}
		}
		ht[pos] = newEntry;
		return no_of_comp;
	}
	void Search(int key) {
		int hash_val = getHashVal(key);
		int pos = hash_val;
		while (ht[pos].key != -1 && ht[pos].key != key) {
			pos = (pos+1) % MAX_SIZE;
			if (pos == hash_val)
				break;
		}

		if (ht[pos].key == key) {
			cout << "Entry is found.\n";
			ht[pos].PrintEntry();
		}
		else
			cout << "Entry is not found.\n";
	}
	void Delete(int key) {
		int hash_val = getHashVal(key);
		int pos = hash_val;
		while (ht[pos].key != -1 && ht[pos].key != key) {
			pos = (pos+1) % MAX_SIZE;
			if (pos == hash_val)
				break;
		}

		if (ht[pos].key == key) {

			int next = (MAX_SIZE+pos-1) % MAX_SIZE;
			while (ht[next].key == -1 || next != hash_val)
				next = (MAX_SIZE+next-1) % MAX_SIZE;

			ht[pos] = ht[next];
			ht[next].key = -1;
		}
		else
			cout << "Entry is not found.\n";
	}
	void PrintHashTable() {
		cout << "Hash Table contains following entries:\n";
		for (int i = 0; i < MAX_SIZE; i++) {
//			if (ht[i].key != -1)
				cout << i << ": ";
				ht[i].PrintEntry();
		}
	}
};

int main() {
	HashTable ht;
	int keys[] = {5, 6, 55, 555, 66, 666};
	string vals[] = {"A", "B", "C", "D", "E", "F"};

	for (int i = 0; i < 6; i++)
		cout << ht.Insert(keys[i], vals[i]) << " ";
	cout << endl;

	ht.PrintHashTable();
	ht.Delete(5);
	ht.PrintHashTable();
	return 0;
}
